File: 백준/Silver/14716. 현수막/현수막.java
import java.io.*;
import java.util.*;

public class Main {
    static int M, N;
    static int[][] graph;
    static int[] dx = {0, 0, 1, 1, 1, -1, -1, -1};
    static int[] dy = {1, -1, 0, -1, 1, 0, -1, 1};

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        M = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());

        graph = new int[M][N];

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }

//        System.out.println(Arrays.deepToString(graph));

        int cnt = 0;
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                if (graph[i][j] == 1) {
                    dfs(i, j);
                    cnt+=1;
                }
            }
        }
        System.out.println(cnt);
    }

    static void dfs(int startX, int startY) {
        Deque<int[]> stack = new ArrayDeque<>();
        stack.push(new int[]{startX, startY});
        graph[startX][startY] = 0; // mark visited

        while (!stack.isEmpty()) {
            int[] cur = stack.pop();
            int x = cur[0]; int y = cur[1];

            for (int i = 0; i < 8; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (!(0 <= nx && nx < M && 0 <= ny && ny < N)) {
                    continue;
                }
                if (graph[nx][ny] == 0) {
                    continue;
                }

                graph[nx][ny] = 0;
                stack.push(new int[]{nx, ny});
            }
        }
    }
}

File: 백준/Silver/14716. 현수막/현수막.py
M, N = map(int, input().split())

graph = [list(map(int, input().split())) for _ in range(M)]

dx = [0, 0, 1, 1, 1, -1, -1, -1]
dy = [1, -1, 0, -1, 1, 0, -1, 1]


def dfs(start_x, start_y):
    st = [(start_x, start_y)]

    while st:
        x, y = st.pop()

        for i in range(8):
            nx = x + dx[i]
            ny = y + dy[i]

            if not (0 <= nx < M and 0 <= ny < N):
                continue
            if graph[nx][ny] == 0:
                continue

            graph[nx][ny] = 0
            st.append((nx, ny))



cnt = 0
for i in range(M):
    for j in range(N):
        if graph[i][j] == 1:
            dfs(i, j)
            cnt += 1
print(cnt)
